catroid/src/main/java/org/catrobat/catroid/formulaeditor/InternToExternGenerator.java

```diff
@@ -319,6 +310,54 @@ private String generateExternStringFromToken(InternToken internToken, boolean tr
 		}
 	}
 
+	private String generateExternNumberStringForToken(String number, boolean trimNumbers) {
+
+		if (trimNumbers) {
+			number = getDesiredStringRepresentattion(number);
+		}
+
+		if (!number.contains(".")) {
+			return number;
+		}
+
+		String left = number.substring(0, number.indexOf('.'));
+		String right = number.substring(number.indexOf('.') + 1);
+
+		return left + getExternStringForInternTokenValue(".", context) + right;
+	}
+
+	private String getDesiredStringRepresentattion(String number) {
+		Float convertedNumber = Float.parseFloat(number);
+		number = convertedNumber.toString();
+
+		if (checkIfExponentExists(number)) {
+			number = cutTrailingZerosofString(number);
+		}
+
+		return number;
+	}
+
+	private boolean checkIfExponentExists(String number) {
+		if (!number.contains("E")) {
+			return true;
+		}
+
+		return false;
+	}
+
+	private String cutTrailingZerosofString(String number) {
+
+		BigDecimal decimal = new BigDecimal(number);
+
+		if (decimal.compareTo(BigDecimal.ZERO) == 0) {
+			decimal = BigDecimal.ZERO;
```

Comment:
Please explain.
If it's because of negative zero, please provide a code snippet where the result is `-0`.

Reply:
> Hmm, why did you change the InternToExternGenerator.java file?

Since all the things which have to do with value showing are done there, i thought i should also do this things there to have a good overview.

Actually the code snippet has only to do with the faulty implementation of BigDecimal (stripTrailingZeros()) in combination with 0. Zero will always be given back as 0.0 even tough this function main purpose is to cut the trailing zeros. In order to handle the 0 also correctly this is there.

In the espresso tests i've seen a test fails with my implementation because i'm also handling really big numbers there with the exponend. Since i was a bit confused I had a look at the current version how it is dealt there and I was a bit shocked. The current implementation does not behave as it should.

Current Problem:
Try to insert 123456789 in any field eg WaitBrick. After you press ok in the FormulaEditor 123456792 is displayed, but in the background still everything is computed with the correct value. But what would we expect to see either 123456789 or with the exponent as for the small numbers as it probably was intendet (1.2345679E8).

Reply:
Thanks for the update!

> decimal.compareTo(BigDecimal.ZERO) == 0

Nice catch! In such cases, a code comment is always appreciated/appropriate :wink:

> 123456789

If you remove `FormulaBrick.getStringForPrototypeView`, restore `Utils.getNumberStringForBricks` but replace it with your `FormulaBrick.getStringForPrototypeView` code, then the `FormulaBrick` and `InternToExternGenerator` would us the same code. The `InternToExternGenerator` would still handle localization too.
What do you think? Please let me know, I still have another request :scream:

Reply:
yeah sure, I'll make the changes tomorrow

Reply:
Awesome! Thanks!

> I still have another request :scream:

Ah, maybe next time :wink:
I would've just suggested to split your changes into 2 commits ([CAT-89](https://jira.catrob.at/browse/CAT-89) and [CAT-2182](https://jira.catrob.at/browse/CAT-2182)). Maybe next time :smile_cat:

Reply:
we can do it that way if you prefer it